traffic_server/plugins/astats_over_http/Makefile.tsxs

```diff
@@ -0,0 +1,27 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+TSXS?=tsxs
+PLUGIN=astats_over_http
+
+all:	$(PLUGIN).c
```

Comment:
`make` is for exactly what it sounds like: making things. When you write a rule like that, you're saying to `make` "This is a recipe that makes a file named 'all'. It requires `$(PLUGIN).c` and these are the steps in the recipe:"

When you make a rule like this, you're sort of losing the major advantages of a Makefile. Since this recipe only creates the plugin file, you should tell `make` the truth - that this recipe creates the plugin.

```make
# You could also do `$(shell which tsxs)` if you're worried about it not being in `$PATH`
TSXS?=tsxs

# `:=` sets the variable for calculation once - syntactically meaningless for a constant,
# but for things like `SOURCE` that *aren't* constant it calculates it once as opposed to
# every time the variable is expanded.
PLUGIN := astats_over_http.so
SOURCE := $(PLUGIN:.so=.c) #name substitution shortcut - technically GNU-only I think

# '$@' expands to the recipe's target; as for '$<' you can use '$?' instead in this case,
# I just prefer '$<'.
$(PLUGIN): $(SOURCE)
    $(TSXS) -v -o $@ $<
```

Now, obviously you still want to be able to do `make && make install` and `make clean`. The recipe that gets run when you call `make` with no arguments is simply the first one specified in the file; `all` has no special meaning to `make`. So already you've got one out of three things. Of course, the install process doesn't create a local file named `install`, but you still want to be able to do `make install`. When a target of a rule isn't created by that rule, you should add it as a dependency of the special `.PHONY` target like this:

```make
.PHONY: install clean

install: $(PLUGIN)
    $(TSXS) -i -o $<

# Using `$(RM)` is recommended for POSIX compliance - it will always
# expand to the path to a binary for removing files forcibly (typically `/bin/rm -f`)
clean:
    $(RM) $(PLUGIN)

# Explicitly removing `$(PLUGIN)` avoids destroying objects people may have
# created from other sources.
```

Reply:
env TSXS=~/opt/ats8/bin/tsxs make -f Makefile.tsxs
env TSXS=~/opt/ats7/bin/tsxs make -f Makefile.tsxs

This make file is a ripoff of other ones with a slight change.  You could argue that one could just invoke it via command line directly

Reply:
I'm not really sure what to make of those first two lines. Of course anything that can be produced by a Makefile could also be produced with the recipe's direct command line steps, but if a Makefile is provided it should be a good one.